dns_sd_bonjour.c

```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
+							  address_v4, sizeof(address_v4))) {
+					have_v4 = TRUE;
 				}
 			case AF_INET6:
 				if (inet_ntop(sa->sin_family, &sa->sin_addr,
-							  dd->address_v6, sizeof(dd->address_v6))) {
-					dd->have_v6 = TRUE;
+							  address_v6, sizeof(address_v6))) {
+					have_v6 = TRUE;
 				}
 		}
 	}
 
+	if (!have_v4 && !have_v6) {
+		WARNING("DNS SD: Can't resolve valid address for service %s.\n", name);
+		goto exit;
+	}
+
+	/* Set properties on the last element on the list. */
+	while (dd->next)
+		dd = dd->next;
+
+	dd->port = port;
+	dd->hostname = strdup(hostname);
+	if (have_v4) {
+		strncpy(dd->addr_str, address_v4, sizeof(dd->addr_str));
+	} else if(have_v6) {
+		strncpy(dd->addr_str, address_v6, sizeof(dd->addr_str));
+	} else {
+		 DEBUG("DNS SD: no ip for %s\n", hostname);
+	}
+
+	DEBUG("DNS SD: added %s (%s:%d)\n", hostname, dd->addr_str, port);
+
+	if (have_v4 || have_v6) {
+		// A list entry was filled, prepare new item on the list.
+		if (!new_discovery_data(&dd->next)) {
+			/* duplicate lock */
+			dd->next->lock = dd->lock;
+		} else {
+			ERROR("DNS SD Bonjour Resolver : memory failure\n");
+		}
+	}
+
+verify_flags:
 	if ((flags & kCFNetServiceFlagMoreComing) == 0) {
 		DEBUG("DNS SD: No more entries coming.\n");
-		goto stop_browsing;
+		CFNetServiceBrowserStopSearch(browser, &anError);
 	}
-	
+
+exit:
+	iio_mutex_unlock(dd->lock);
 	return;
 
 stop_browsing:
-	
 	CFNetServiceBrowserStopSearch(browser, &anError);
 }
 
-int discover_host(char *addr_str, size_t addr_len, uint16_t *port)
+int dnssd_find_hosts(struct dns_sd_discovery_data ** ddata)
 {
 	int ret = 0;
-	
-	struct cfnet_discovery_data ddata = { FALSE, FALSE };
-	CFNetServiceClientContext clientContext = { 0, &ddata, NULL, NULL, NULL };
-	CFStreamError error;
-	Boolean result;
+	struct dns_sd_discovery_data *d;
 
-	CFStringRef type = CFSTR("_iio._tcp.");
-	CFStringRef domain = CFSTR("");
+	DEBUG("DNS SD: Start service discovery.\n");
 
-	DEBUG("DNS SD: Resolving hostname.");
+	if (new_discovery_data(&d) < 0) {
+		return -ENOMEM;
+	}
+
+	d->lock = iio_mutex_create();
+	if (!d->lock) {
+		dnssd_free_all_discovery_data(d);
+		return -ENOMEM;
+	}
 
-	CFNetServiceBrowserRef gServiceBrowserRef = CFNetServiceBrowserCreate(
+	CFNetServiceClientContext clientContext = { 0, d, NULL, NULL, NULL };
+	CFNetServiceBrowserRef serviceBrowser = CFNetServiceBrowserCreate(
 		kCFAllocatorDefault, __cfnet_browser_cb, &clientContext);
 
-	if (gServiceBrowserRef == NULL) {
+	if (serviceBrowser == NULL) {
 		ERROR("DNS SD: Failed to create service browser.\n");
-		ret = ENOMEM;
+		dnssd_free_all_discovery_data(d);
+		ret = -ENOMEM;
 		goto exit;
 	}
-	result = CFNetServiceBrowserSearchForServices(
-		gServiceBrowserRef, domain, type, &error);
+
+	CFRunLoopRef runLoop = CFRunLoopGetCurrent();
+	CFNetServiceBrowserScheduleWithRunLoop(serviceBrowser, runLoop, kCFRunLoopDefaultMode);
+
+	CFStringRef type = CFSTR("_iio._tcp.");
+	CFStringRef domain = CFSTR("local.");
```

Comment:
I think that domain shall be empty string to be consistent with Avahi implementation. It is stated on [Apple's page](https://developer.apple.com/library/archive/documentation/Networking/Conceptual/NSNetServiceProgGuide/Articles/CFNetServices.html) that:

> Domain. The domain for the service, typically the empty string (CFSTR(""))for default domain(s), or local. for the local domain only

Reply:
and since we want the local domain only - isn't this what we _do_ want?

Reply:
I compared with the Avahi implementation where NULL is passed for domain and[ Avahi documentation](https://www.avahi.org/doxygen/html/lookup_8h.html#a52d55a5156a7943012d03e6700880d2b) says:

> avahi_service_browser_new()
> 
> domain | A domain to browse in. In most cases you want to pass NULL here for the default domain (usually ".local")

That sounds to me the same like the passing `CFSTR("")` on macOS.

Reply:
Ok - I updated to a empty string.